patch_rebaser/patch_rebaser.py

```diff
@@ -109,6 +111,96 @@ def set_up_git_config(name, email):
     os.environ["GIT_COMMITTER_EMAIL"] = email
 
 
+class Rebaser(object):
+
+    def __init__(self, repo, branch, commit, remote, tag_name,
+                 dev_mode=True, max_retries=3):
+        """Initialise the Rebaser
+
+       :param git_wrapper.GitRepo repo: An initialised GitWrapper repo
+       :param str branch: Local branch name to rebase
+       :param str commit: Commit sha to rebase to
+       :param str remote: Remote name
+       :param str tag_name: Tag name to push to the remote
+       :param bool dev_mode: Whether to run the push commands as dry-run only
+       :param int max_retries: How many retry attempts if remote changed during
+                               rebase
+        """
+        self.repo = repo
+        self.branch = branch
+        self.commit = commit
+        self.remote = remote
+        self.tag_name = tag_name
+        self.dev_mode = dev_mode
+        self.max_retries = max_retries
+
+        self.remote_branch = "{0}/{1}".format(self.remote, self.branch)
+
+    def rebase_and_update_remote(self):
+        """Rebase the local branch to the specific commit & push the result."""
+        self.repo.remote.fetch(self.remote)
+
+        # Reset the local branch to the latest
+        self.repo.branch.create(self.branch,
+                                self.remote_branch,
+                                reset_if_exists=True)
+
+        # Tag the previous branch tip, before rebase
+        tag = self.repo.repo.create_tag(self.tag_name, self.remote_branch)
+
+        # Rebase
+        self.perform_rebase()
+
+        # Check if any new changes have come in
+        self.repo.remote.fetch(self.remote)
+
+        # TODO: Update git_wrapper to offer a more sensible comparison function
+        remote_tip_ref = self.repo.repo.remotes[self.remote].refs[self.branch]
+        remote_sha = remote_tip_ref.commit.hexsha
+        if tag.commit.hexsha != remote_sha:
+            if self.max_retries > 0:
+                LOGGER.info("Remote changed during rebase. Remaining "
+                            "attempts: %s", self.max_retries)
+                time.sleep(20)
+                self.max_retries -= 1
+
+                # We'll need to move the tag to the new HEAD
+                self.repo.git.tag("-d", self.tag_name)
+                self.rebase_and_update_remote()
+            else:
+                LOGGER.warning("Remote changed multiple times during rebase, "
+                               "not pushing.")
```

Comment:
Note: this means the rebase result will still get built, so the patches branch will temporarily be out of date with the build. This seems more desirable than resetting to the remote branch and skipping having an up-to-date build (no latest upstream, etc) this round.

In either case this will be corrected the next time the script is run.